### ISSUE TITLE
Ergonomic multi-block operations

### ISSUE THREAD

**Issue by kianenigma** [paritytech/substrate#9587]
I have started writing multi-block code in a few occasions already. There are multiple ways to abstract it, but this is the ultimate one:

```rust 
fn multi_block_function(n: T::BlockNumber) -> Weight {
	// this only executes once. 
	let pre_process = foo(); 
		
	// will generate some storage items that stores the current block number,
	// with some match statements that re-executes this code for the coming 10 blocks. 
	#[pallet::multi_block]
    for i in 0..10 {
		// this is executed 10 times, over the course of 10 blocks
		// variable `n` should be shadowed to the correct value here.
		let _ = bar(n)?;
	} 
	
	// this is exeucted once, right after the last iteration.
	let post_process = baz();
}
```

This is merely an idea at this stage and there are a lot of caveats, which should be discussed in depth before an attempt at implementation. One I have more concrete ideas about this I will move it to the frame vision project.

**Comment by kianenigma** [comment]
@Ank4n

**Comment by xlc** [comment]
What we really want is async operations. We already have proper abstraction for it with years of experiences and I want to make sure we don't repeat the same mistakes and introduce more race condition bugs

**Comment by ggwpez** [comment]
> What we really want is async operations.

What kind of operations? For example iterators on `StorageMaps`? So that `StorageMap::iter_keys().map(` would be multi-block operation.  
Or rather wrapping a whole migration code block in some kind of async macro and using storage operations as yield points?  
```rust
multi_block! {
    // migration code that will be split into multiple blocks.
}
```
I think this is more complicated than having some kind of block number counter.  
What are the race conditions do you see with this approach?

**Comment by xlc** [comment]
The main reason we want to split some operations to multiple blocks is due weight limit, not just because we want to do it due to business reason (which I think is out of scope for this issue). 

Before a language natively supports async operations, we use callbacks to implement it. So it will be something like this

```rs
fn do_work(i: u32) {
  // do something with i
}

fn get_work(): Option<u32> {
 // return next work
}

fn get_and_do_work() -> Weight {
  if let Some(work) = get_work() {
    do_work(work)
     multi_block_helper::execute_or_queue(Call::get_and_do_work)
    return weight_info::get_and_do_work(work)
  }
  return weight_info::get_work()
}

fn migration() {
   multi_block_helper::execute_or_queue(Call::get_and_do_work)
}
```

Later, if we somehow get async support, we can make it

```rs

async fn get_and_do_work() -> Weight {
  if let Some(work) = get_work() {
    do_work(work)
    get_and_do_work().await
    return weight_info::get_and_do_work(work)
  }
  return weight_info::get_work()
}

async fn migration() {
    get_and_do_work().await
   // code here will be executed after all the work is completed
}
```

**Comment by ggwpez** [comment]
Okay so there are a few other cases where multi-block or multi-call aka. *async* operations would be usefull:  
https://github.com/paritytech/substrate/pull/12367#discussion_r991535381 (which uses `remove_prefix`)
https://github.com/paritytech/substrate/pull/12310 (which also wants to delete a `StorageMap` which could produce a too high weight for one call/block)

**Comment by bkchr** [comment]
> Maybe we can write the changes of a migration into a new storage transaction (aka changeset) and commit that once its done. That would split the reading part (PoV intense) into multiple blocks and put the writing part into the last block. Ideally it just swaps two sub-tree hashes and does not copy anything.

This would not work. How should that work for Parachains? The validation is always stateless. The changeset would also not be tracked as part of the state.

**Comment by kianenigma** [comment]
The original idea of this issue is being overlooked a bit, which is tailored towards providing a easy to use syntax to write multi-block operations, much like [OpenMP's directives](https://www.ibm.com/docs/en/xl-c-aix/13.1.0?topic=programs-using-openmp-directives) or [Rust's Rayon macros](https://github.com/rayon-rs/rayon-macro). This certainly won't be generalizable. If you want a generalizable, writing it from scratch using on-idle or on-initialize (as done by Acala's scheduler) is not super hard.

**Comment by gavofyork** [comment]
FWIW Async operations would also be a gamechanger for XCM/XCMP. Right now the best we can do is have a dispatchable be called when an XCM reply arrives. Far far better is if we could so a multi-block `await` on the reply.

**Comment by gavofyork** [comment]
I think it's realistic to be able to *mostly* suspend the chain while a multi-block migration completes. Not for every pallet (session/system/staking/babe/grandpa come to mind as non-suspendable), but almost all of the others should be fine to pause without risking bricking the chain.

**Comment by bkchr** [comment]
> I think it's realistic to be able to _mostly_ suspend the chain while a multi-block migration completes. Not for every pallet (session/system/staking/babe/grandpa come to mind as non-suspendable), but almost all of the others should be fine to pause without risking bricking the chain.

Yes. I also have come to a similar conclusion. Proper multi block migration should involve migrating consensus related storage items in the first block that uses the new runtime and then all the other migrations can happen while the chain is paused.

**Comment by ruseinov** [comment]
> Yes. I also have come to a similar conclusion. Proper multi block migration should involve migrating consensus related storage items in the first block that uses the new runtime and then all the other migrations can happen while the chain is paused.

Sounds good. In the example that I mentioned before the situation with people being able to write to that particular storage absolutely does not work, and that's considering everything is happening on the relay chain only. 
If the chain was paused (no storage i/o) - that migration would be a piece of cake to write and execute. Right now it requires lots of modifications that just block certain extrinsics depending on a storage flag. It's not ideal and very error-prone.

**Comment by bkchr** [comment]
> Right now it requires lots of modifications that just block certain extrinsics depending on a storage flag. It's not ideal and very error-prone.

We would block all extrinsics beside the inherents. Migrations will not run for that long time, so it is fine to not include extrinsics for a short period of time.

**Comment by ruseinov** [comment]
Would be great to see this going forward. Happy to lend a hand where I can, sounds like an interesting problem to tackle with a lot to gain.

I'm assuming there isn't a good way to "pause" pallets now. As far as I understand that would require some sort of flag that would switch things off under the hood until a certain cached storage item gets set/unset. 
Since we can't really pause the storage due to the operations on it being infallible, I guess that means some sort of a global flag that could be disabled per pallet for the pallets mentioned above that can't be fully paused.

**Comment by ggwpez** [comment]
>Since we can't really pause the storage due to the operations on it being infallible, I guess that means some sort of a global flag that could be disabled per pallet for the pallets mentioned above that can't be fully paused.

https://github.com/paritytech/substrate/pull/12092 could be used I think. Or maybe we need some more fundamental pausing of all extrinsics, not sure.

**Comment by bkchr** [comment]
> #12092 could be used I think. Or maybe we need some more fundamental pausing of all extrinsics, not sure.

No this should be exactly what we need/want.

**Comment by ruseinov** [comment]
> > Since we can't really pause the storage due to the operations on it being infallible, I guess that means some sort of a global flag that could be disabled per pallet for the pallets mentioned above that can't be fully paused.
> 
> #12092 could be used I think. Or maybe we need some more fundamental pausing of all extrinsics, not sure.

Looks sufficient to me too.
